src/lc3/vm.rs

```diff
@@ -0,0 +1,140 @@
+use std::fs;
+
+const MEMORY_MAX: usize = 1 << 16;
+pub struct VM {
+    r0: u16,
+    r1: u16,
+    r2: u16,
+    r3: u16,
+    r4: u16,
+    r5: u16,
+    r6: u16,
+    r7: u16,
+    pc: u16,
+    cond: ConditionFlag,
+    running: bool,
+    memory: [u16; MEMORY_MAX],
+}
+
+#[derive(Debug)]
+pub enum VMError {
+    ReadingFile(String),
+    ConcatenatingBytes(String),
+    Adding,
+}
+
+enum Opcode {
+    BR,   /* branch */
+    Add,  /* add  */
+    LD,   /* load */
+    ST,   /* store */
+    Jsr,  /* jump register */
+    And,  /* bitwise and */
+    Ldr,  /* load register */
+    Str,  /* store register */
+    Rti,  /* unused */
+    Not,  /* bitwise not */
+    Ldi,  /* load indirect */
+    Sti,  /* store indirect */
+    Jmp,  /* jump */
+    Res,  /* reserved (unused) */
+    Lea,  /* load effective address */
+    Trap, /* execute trap */
+}
+
+enum ConditionFlag {
+    Pos, // = 1 << 0, /* P */
+    Zro, // = 1 << 1, /* Z */
+    Neg, // = 1 << 2, /* N */
+}
+
+impl VM {
+    pub fn new() -> VM {
+        VM {
+            r0: 0,
+            r1: 0,
+            r2: 0,
+            r3: 0,
+            r4: 0,
+            r5: 0,
+            r6: 0,
+            r7: 0,
+            pc: 0x3000,
+            cond: ConditionFlag::Zro,
+            running: false,
+            memory: [0; MEMORY_MAX],
+        }
+    }
+
+    pub fn read_image(&mut self, image_path: &str) -> Result<(), VMError> {
+        let content = &fs::read(image_path)
+            .map_err(|_| VMError::ReadingFile(String::from("Error on reading from file path")))?;
```

Comment:
We could return the file-path for better context when debuggin?
```suggestion
            .map_err(|e| VMError::ReadingFile(format!("Failed to read file {}: {}", image_path, e))?;
```

Reply:
Nice! great suggestion